feat(atc): support named var.

evanchaoli opened the pull request:
# Existing Issue

This is a a follow-up enhancement of PR #4600, and complies with RFC [Conourse/RFC#39](https://github.com/concourse/rfcs/pull/39).

# Why do we need this PR?

This PR makes pipeline var_source fully functional.

# Changes proposed in this pull request

Basically this change is neat.

* The key part is to add a new struct `namedSecrets`. It extract var_source name from `secretPath`, then checks var_source and forward real `Get` to the underlying secrets.
* `VarSourcePool` will always returns `namedSecrets` with original secrets wrapped in it.
* The regexp is updated in `template.go` to accept named var like `((vs:var))`.
* Unit tests and testflight have been updated.

# Contributor Checklist
> Are the following items included as part of this PR? Please delete checkbox items that don't apply.
- [x] Unit tests
- [x] Integration tests (if applicable)
- [ ] Updated documentation (located at https://github.com/concourse/docs)
- [ ] Updated release notes (located at https://github.com/concourse/concourse/tree/master/release-notes)


# Reviewer Checklist
> This section is intended for the core maintainers only, to track review progress. Please do not
> fill out this section.
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the [BOSH](https://github.com/concourse/concourse-bosh-release) 
      and [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for the [integration tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env) (for example, if they are Garden configs that are not displayed in the `--help` text).

evanchaoli commented:
My test pipeline:

```yml
var_sources:
- name: dev-vault
  type: vault
  config:
    url: http://192.168.1.12:8200
    path_prefix: concourse
    client_token: ((my-vault-token))

resource_types:
- name: rt-tiny
  type: registry-image
  source:
    repository: evanli/tiny2
    tag: v2

resources:
- name: tiny
  type: rt-tiny
  check_every: 2m
  source:
    key: ((dev-vault:hello))

jobs:
- name: test
  public: true
  plan:
  - get: tiny
    params:
      other_key: ((dev-vault:test.pass))

  - task: task-should-succeed
    config:
      platform: linux
      image_resource:
        type: registry-image
        source: {repository: busybox}
      run:
        path: sh
        args:
         - -exc
         - |
           echo The name is ((dev-vault:test.pass))

  - task: task-should-fail
    config:
      platform: linux
      image_resource:
        type: registry-image
        source: {repository: busybox}
      run:
        path: sh
        args:
         - -exc
         - |
           echo The name is ((test.pass))
```

Where the first task use var_source name with the vars, which should pass; and the second task uses var names without var_source name, then it will look for vars from the system vault, and will not find, then fail.

The UI screen shot proves the test passed:

![named_var_source](https://user-images.githubusercontent.com/18585861/69022119-317af280-09f5-11ea-9ac6-42cf98864f9c.png)

taylorsilva commented:
Did you mean to reference issue #4579 instead of #4580?

evanchaoli commented:
@tylorsilva sorry for the confusion. I think 4580 was added by auto completion. I didn’t notice that, just removed that.

evanchaoli commented:
@vito Can you please review this? It’s ready for review.

evanchaoli commented:
I added a new config `skip_team_path` to Vault, so that in `var_sources`, secret lookup may skip team paths. This behavior complies with [RFC var_sources](https://github.com/concourse/rfcs/pull/39).

evanchaoli commented:
@vito I communicated with some of our users this week, and they felt that allowing a var_source to refer to previous var_sources would be helpful, which is same as what you have proposed in [RFC: var_sources](https://github.com/concourse/rfcs/pull/39). So I just added a new commit to support that, as you can see, the code is actually very simple.

evanchaoli commented:
Just rebased and resolved conflicts.

vito commented on atc/db/pipeline.go:
```diff
@@ -1136,35 +1138,40 @@ func (p *pipeline) getBuildsFrom(tx Tx, col string) (map[string]Build, error) {
 	}
 
-	varss = append(varss, globalVars)
 	return vars.NewMultiVars(varss), nil
```
What if instead of just listing everything under `vars.NewMultiVars` and having var sources know their own name, we could have our own `vars.NamedVariables` type which does the lookup by-name through a map?

Something like this:

```go
namedVars := vars.NamedVariables{}

allVars := vars.NewMultiVars([]vars.Variables{globalVars, namedVars})

for _, cm := range orderedVarSources {
  // ...

  // allVars here sneakily relies on us mutating namedVars
  // ¯\_(ツ)_/¯ should be safe!
  newConfig, err := creds.NewParams(allVars, atc.Params{"config": cm.Config}).Evaluate()
  if err != nil {
    // ...
  }

  // this doesn't need a name anymore
  secrets, err := varSourcePool.FindOrCreate(logger, config, factory)
  if err != nil {
    // ...
  }

  // ...

  namedVars[cm.Name] = creds.NewVariables(secrets, p.TeamName(), p.Name(), true)
}

return allVars, nil
```

Then `(vars.NamedVariables).Get` type would just do a map lookup and delegate that way.

evanchaoli replied:
This is a good idea, `vars.NamedVariables` is betters than `creds.NamedSecrets`.

evanchaoli commented on atc/config_test.go:
```diff
@@ -40,4 +40,107 @@ var _ = Describe("Config", func() {
 		})
 	})
+
+	Describe("VarSourceConfigs.OrderByDependency", func() {
```
Thanks.

evanchaoli commented on atc/creds/vault/manager.go:
```diff
@@ -19,6 +19,7 @@ type VaultManager struct {
 	PathPrefix string `long:"path-prefix" default:"/concourse" description:"Path under which to namespace credential lookup."`
 	SharedPath string `long:"shared-path" description:"Path under which to lookup shared credentials."`
 	Namespace  string `long:"namespace"   description:"Vault namespace to use for authentication and secret lookup."`
+	SkipTeamPath bool `long:"skip-team-path" description:"Skip team/ and team/pipeline/ from lookup path"`
```
I'm ok. Fortunately `SkipTeamPath` was added with a standalone commit, so it's easy to revert it, and get it back in a separate PR.

evanchaoli commented on atc/creds/pool.go:
```diff
@@ -43,13 +44,13 @@ func (pool *varSourcePool) Size() int {
 	}
 
-	key := string(b)
+	key := varSourceName + string(b)
```
As we add `vars.NamedVariables`, `pool` doesn't need to realize var source name. I'll remove varSourceName from pool.

evanchaoli commented on atc/configvalidate/validate_test.go:
```diff
@@ -219,6 +219,29 @@ var _ = Describe("ValidateConfig", func() {
+				Expect(errorMessages).To(HaveLen(1))
+				Expect(errorMessages[0]).To(ContainSubstring("duplicate var_source name: some"))
+			})
+		})
```
Make sense. Will add.

evanchaoli commented:
@vito I have just addressed all review comments:

* Removed `creds.NamedSecrets` and anded `vars.NamedVariables`.
* Restored `pool.go`, so that `pool.go` is not touched in this PR. With `vars.NamedVariables`, pool doesn't need to realize var source name at all.
* Added test cases to config validate.
* Reverted `skipTeamPath`.

I hope var source to be included in 5.8, because our users are waiting for that, I don't want them to wait the other month for 5.9. Also, I plan to work on input_vars, based on code of this PR, input_var should be easy to implement. 

My roadmap is to contribute "input_vars" and "OPA integration" to 5.9, I have submitted RFCs for both of them. "OPA integration" is an urgent feature for us, actually I have finished coding but haven't submitted a PR yet. "input_vars" is not urgent, but very nice to have, that will somewhat change the way of creating pipelines, reducing development effort of tasks and resource types.

vito commented on vars/named_vars.go:
```diff
@@ -0,0 +1,54 @@
+	"strings"
+)
+
+type NamedVariables struct {
```
This could actually be a type alias:

```go
type NamedVariables map[string]Variables
```

And then it can be constructed like `vars.NamedVariables{}` instead of constructing a map. And then `NewNamedVariables` could be removed.

vito commented on atc/db/pipeline.go:
```diff
@@ -1136,36 +1138,46 @@ func (p *pipeline) getBuildsFrom(tx Tx, col string) (map[string]Build, error) {
-	varss := []vars.Variables{}
-	for _, cm := range p.varSources {
+	namedVarsMap := map[string]vars.Variables{}
+	// It's safe to construct NewNamedVariables with namedVars here, because
```
:+1: worth commenting since it's a little tricky

vito commented:
Looks like `testflight` is failing. Here are some relevant logs:

```json
web_1     | {"timestamp":"2019-12-11T17:51:26.763230340Z","level":"debug","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.found-existing-credential-manager","data":{"pipeline":"tf-pipeline-4-5a94a3bd-e9d4-4343-6453-6ee8ddabdf3c","resource":"resource-using-vars","session":"16.55.27.1.1","team":"testflight"}}
web_1     | {"timestamp":"2019-12-11T17:51:26.763466054Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-evaluate-resource-source","data":{"error":"undefined vars: vs:initial_version_var, vs:some_numeric_var","pipeline":"tf-pipeline-4-5a94a3bd-e9d4-4343-6453-6ee8ddabdf3c","resource":"resource-using-vars","session":"16.55.27.1.1","team":"testflight"}}
web_1     | {"timestamp":"2019-12-11T17:51:26.765630634Z","level":"error","source":"atc","message":"atc.pipelines.radar.failed-to-run-scan-resource","data":{"error":"undefined vars: vs:initial_version_var, vs:some_numeric_var","pipeline":"tf-pipeline-4-5a94a3bd-e9d4-4343-6453-6ee8ddabdf3c","session":"16.55","team":"testflight"}}
```

evanchaoli commented:
> Looks like `testflight` is failing. Here are some relevant logs:
> 

@vito Thank the `testflight` for catching the bug. The bug was introduced yesterday when I changed to `NamedVariable`, which was only hit when there was a single `var_source` defined in a pipeline. While both my local tests and unit tests used multiple `var_sources`, so the bug was not caught.

I have fixed the bug, and enhanced the unit tests to cover both single var_source and multiple var_sources cases.

BTW, why are we no longer able to see build logs? Looks like the pipelines are not exposed to public now.

evanchaoli commented on vars/named_vars.go:
```diff
@@ -0,0 +1,47 @@
+		return vars.Get(VariableDefinition{Name: varName})
+	}
+
+	return nil, false, nil
```
Make sense.

evanchaoli replied:
@vito As I'm working on `input_vars` today, and I rethink this over. The problem I'm facing is, `input_vars` are not secrets, if we merge `input_vars`'s `NamedVars` into `var_sources`'s `NamedVars`, then `input_vars` would be redacted, which I think is not intended behavior.

For your concern, since `var_sources` are not allowed to have duplicate names, as long as we make sure `input_vars` don't use duplicate names as `var_sources`, then problem will not happen. In the future, for project, we may also check `var_sources` within a project to not have duplicate names. WDYT?

As 5.8.0 is not release yet, if you agree, now is the best time to change it back to return no error.

vito replied:
Sorry, I don't follow - how does returning an error here impact this decision?

Redacting `input_vars` *could* be what the user wants - it really depends on the contents of the vars. Someone could have generated a username/password to a temporary account, for example. Maybe the decision to redact should be made fine-grained, i.e. an attribute of the returned var, and things like `input_vars` would allow you to configure it?

evanchaoli replied:
Returning an error here make a step to have only one `NamedVars`. `input_vars` has to be merged into `var_sources`. If here doesn't return error, then we may have multiple `NamedVars`, one for `var_sources` and the other for `input_vars`.

As `input_vars` is a simplified alternative of input files, and content of input files are not redacted, so I'm thinking maybe not redact `input_vars`.

But I don't have strong opinion on that. Let's keep the current way here to return an error.

vito replied:
Ah, I see. Let's leave it as-is for now and hash this detail out in the `input_vars` RFC (...once I have time to take a look at it :sweat:).
